db/compaction/compaction.cc

```diff
@@ -566,6 +567,47 @@ std::unique_ptr<CompactionFilter> Compaction::CreateCompactionFilter(
       context);
 }
 
+std::vector<SstPartitioner::Segment> Compaction::CreateSegmentsForLevel(
+    int in_level) const {
+  // So... the below files should be adjacently sorted.
+  // For now, this is only for creating the next-of-output level info, so it
+  // makes sense for not supporting L0.
+  assert(in_level != 0);
+
+  // Some of test cases may not initialize the version...
+  if (input_version_ == nullptr) {
+    return std::vector<SstPartitioner::Segment>();
+  }
+
+  const auto vsi = input_version_->storage_info();
+  if (in_level >= vsi->num_non_empty_levels()) {
+    // The level shall be empty.
+    return std::vector<SstPartitioner::Segment>();
+  }
+  const auto files = vsi->LevelFilesBrief(in_level);
+  const auto cmp = immutable_options()->user_comparator;
+  const auto bgn = std::lower_bound(
+      files.files, files.files + files.num_files, smallest_user_key_,
+      [cmp](FdWithKeyRange& fd, const Slice& slice) {
+        return cmp->Compare(fd.largest_key, slice) < 0;
+      });
+  const auto end = files.files + files.num_files;
+
+  if (bgn == end) {
+    return std::vector<SstPartitioner::Segment>();
+  }
+
+  std::vector<SstPartitioner::Segment> segs;
+  segs.emplace_back(0, bgn->smallest_key);
+  for (const FdWithKeyRange* iter = bgn; iter < end; iter++) {
+    if (cmp->Compare(iter->smallest_key, largest_user_key_) > 0) {
+      break;
+    }
+    segs.emplace_back(iter->fd.GetFileSize(), iter->largest_key);
```

Comment:
I don't know why test work, but this doesn't copy string, the string will be destructed once iter is out of scope.

Reply:
I'm not sure how to own a `Slice`, it looks like a pure reference type(The content is a `const char*`)... Should we use `PinSlice` here (Or are there other types presents a owned slice?)?

For this case, I'm guessing the slice should live as long as `VersionStorageInfo`. I'm not sure whether that is true, if I got wrong, perhaps it is needed to use a owned type.

Reply:
You are probably right.

Reply:
You are probably right.